src/containers/LoginRegister/index.jsx

```diff
@@ -0,0 +1,201 @@
+import React, { useContext, useState } from "react";
+import { withRouter } from "react-router-dom";
+import { useTranslation } from "react-i18next";
+import { Alert, Spin } from "antd";
+import firebase from "firebase";
+import Login from "../../components/userauth/Login";
+import Register from "../../components/userauth/Register";
+import ResetPassword from "../../components/userauth/ResetPassword";
+import firestore, { auth } from "../../firebaseConfig";
+import "./index.css";
+import kids from "../../images/LoginKids.svg";
+
+const LoginRegisterPage = ({ history }) => {
+  const [t] = useTranslation();
+  const [active, setActive] = useState("login");
+  const [error, setError] = useState("");
+  const [message, setMessage] = useState();
+  const [loading, setLoading] = useState(false);
+
+  const resetErrorAndMessage = () => {
+    setError("");
+    setMessage(<></>);
+  };
+  const login = async (email, password) => {
+    resetErrorAndMessage();
+    try {
+      setLoading(true);
+      await auth.signInWithEmailAndPassword(email, password);
+      setLoading(false);
+      history.push("/");
+    } catch (e) {
+      setLoading(false);
+      setError(e.message);
+    }
+  };
+  const passwordReset = async (email) => {
+    resetErrorAndMessage();
+    try {
+      setLoading(true);
+      await auth.sendPasswordResetEmail(email);
+      setLoading(false);
+      setMessage(
+        <Alert
+          className="authAlert"
+          showIcon
+          message={
+            <div>
+              {t("loginRegister.anEmailSent")}
+              <b>{email}</b>
+              {t("loginRegister.anEmailSentContain")}
+            </div>
+          }
+          type="success"
+        />
+      );
+      setActive("login");
+    } catch (e) {
+      setLoading(false);
+      setError(e.message);
+    }
+  };
+  const loginWithFacebook = async () => {
+    resetErrorAndMessage();
+    try {
+      setLoading(true);
+      const provider = new firebase.auth.FacebookAuthProvider();
+      provider.setCustomParameters({
+        display: "popup",
+      });
+      setLoading(false);
+      history.push("/");
+    } catch (e) {
+      setLoading(false);
+      setError(e.message);
+    }
+  };
+  const loginWithGoogle = async () => {
```

Comment:
please look at how there is a common pattern in this code, and if i had to change something, say line 73 or 74, i have to change it in like 4 places

Reply:
added function for history pushing and catch

Reply:
i wasn't talking about just those two lines. i was talking about that as an example

everything doe sthis
* calls resetErrorAndMessage
* startsloading
* does some other stuff thats different for each time
* does some cleanup / error stuff

think about how you can factor out ALL that common logic, there shouldnt be 20 try catches everywhere

Reply:
PTAL